src/Variable.cpp

```diff
@@ -200,6 +200,26 @@ VariablePtr Variable::subvariable(Space space, std::string_view baseName, Space
   }
 }
 
+VariablePtr Variable::subvariable(Space space, Space shift) const
+{
+  if(!(space * shift <= space_))
+    throw std::runtime_error("[Variable::subvariable] Invalid space and shift dimension");
+
+  VariablePtr base = superVariable();
+  if(isBasePrimitive())
+  {
+    int start = shift_.rSize() + shift.rSize();
+    std::stringstream ss;
+    ss << base->name() << "[" << start << ":" << start + space.rSize() - 1 << "]";
+    return std::make_shared<Variable>(make_shared_token{}, base, space, ss.str(), shift_ * shift);
+  }
+  else
+  {
+    VariablePtr bp = base->basePrimitive();
+    return dot(std::make_shared<Variable>(make_shared_token{}, bp, space, "", shift_ * shift), derivativeNumber_, true);
```

Comment:
So, if I understand it correctly, this:
- creates the "base" sub-variable corresponding to this derivative sub-variable
- derivate that sub-variable

Why is the name `""` for the base sub-variable in that case? Isn't it returned if I get the base primtive of the variable that was just created?

Isn't it broken for non-euclidean spaces? If I have a variable v on `SO(3) x R(N)` and if I do `dot(v).subspace(R(N), R(6))` then wouldn't I end up with a variable in `R(1) x R(N-1)`? (i.e. the last derivative component in `SO(3)` and all except the last derivative in `R(N)`)

Reply:
> Why is the name "" for the base sub-variable in that case? Isn't it returned if I get the base primtive of the variable that was just created?
I was so focused on getting the derivative name right that I forgot about the primitive. Nice catch.

> Isn't it broken for non-euclidean spaces? If I have a variable v on SO(3) x R(N) and if I do dot(v).subspace(R(N), R(6)) then wouldn't I end up with a variable in R(1) x R(N-1)? (i.e. the last derivative component in SO(3) and all except the last derivative in R(N))
The shift needs to be relative to the space of the supervariable. The example you give would indeed have problem, but since we don't know about the precise nature of the space (it could be R(N)xSO(3) as well, for the same dimensions), we cannot guard against that.
The implementation of subvariables is handling correctly subvariables on manifold, but the user needs to be carefull when creating these subvariables.

There is one piece of codes that's assuming Euclidean variables in VariableCountingVector.cpp. For the use we have it's fine, but I think I need to attack the general case next, to avoid any future surprise.

Reply:
Fixed in a533edd46731861420fe2620693aba67b65d642d